File: 1-D_Dynamic_Programming/Easy/LC_509-Fibonacci_Number/LC_509-Fibonacci_Number.cpp
// #include <bits/stdc++.h>
#include <iostream>
#include <cassert>
#include <vector>

using namespace std;

class Solution {
public:
    // Dynamic programming w/ 2 variables
    // T: O(n), M: O(1)
    int fib(int n) {
        if (n <= 1)
            return n;
        
        // Start from F(2)
        int n0 = 0, n1 = 1;
        for (int num = 2; num <= n; num++) {
            int temp = n0 + n1;
            n0 = n1;
            n1 = temp;
        }
        return n1;
    }

    /*
    // Dynamic programming w/ array approach
    // T: O(n), M: O(n)
    int fib(int n) {
        if (n <= 1)
            return n;
        
        // Start from F(2)
        vector<int> dp = {0, 1};
        for (int num = 2; num <= n; num++) {
            dp.push_back(dp[dp.size() - 2] + dp[dp.size() - 1]);
        }
        return dp[n];
    }
    */
};

int main() {
    Solution sol;
    
    // Ex1
    assert(sol.fib(2) == 1);
    // Ex2
    assert(sol.fib(3) == 2);
    // Ex3
    assert(sol.fib(4) == 3);
}

File: 1-D_Dynamic_Programming/Easy/LC_509-Fibonacci_Number/LC_509-Fibonacci_Number.py
class Solution:
    # Dynamic programming w/ 2 variables approach
    # T: O(n), M: O(1)
    def fib(self, n: int) -> int:
        if n <= 1:
            return n
        
        # Start from 2
        n0, n1 = 0, 1
        for num in range(2, n+1):
            n0, n1 = n1, n0 + n1
        return n1
    
    '''
    # Dynamic programming w/ array approach
    # T: O(n), M: O(n)
    def fib(self, n: int) -> int:
        if n <= 1:
            return n
        
        # Start from 2
        dp = [0, 1]
        for num in range(2, n+1):
            dp.append(dp[-2] + dp[-1])
        return dp[-1]
    '''

sol = Solution()

# Ex1
attempt = sol.fib(2)
assert attempt == 1, "Expected 1, but got: " + attempt
# Ex2
attempt = sol.fib(3)
assert attempt == 2, "Expected 2, but got: " + attempt
# Ex3
attempt = sol.fib(4)
assert attempt == 3, "Expected 3, but got: " + attempt
